pulsar-client/src/main/java/org/apache/pulsar/client/util/ExecutorProvider.java

```diff
@@ -45,12 +46,21 @@ public ExecutorProvider(int numThreads, ThreadFactory threadFactory) {
         for (int i = 0; i < numThreads; i++) {
             executors.add(Executors.newSingleThreadScheduledExecutor(threadFactory));
         }
+        isShutdown = false;
     }
 
     public ExecutorService getExecutor() {
         return executors.get((currentThread.getAndIncrement() & Integer.MAX_VALUE) % numThreads);
     }
 
+    public ExecutorService getExecutor(Object object) {
+        return getExecutor(object == null ? -1 : object.hashCode() & Integer.MAX_VALUE);
+    }
+    
+    public ExecutorService getExecutor(int hash) {
```

Comment:
In general, from API perspective, it would be better to just expose an Object (or String) for selecting the thread. Couldn't we calculate the hash internally always?

Reply:
Ok, we can move the key hash calculate internally and replace this method to getExecutor(byte[]).